File: src/main/java/com/theendercore/hydra/config/ModConfig.java
package com.theendercore.hydra.config;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.fabricmc.loader.api.FabricLoader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;

import static com.theendercore.hydra.HydraMod.MODID;

public class ModConfig {

    public static final String DEFAULT_USERNAME = "";
    public static final String DEFAULT_OAUTH_KEY = "";
    public static final String DEFAULT_PREFIX = "!";
    public static final Color2 DEFAULT_CHANNEL_CHAT_COLOR = Color2.BLUE;
    public static final String DEFAULT_TIME_FORMATTING = "HH:mm";
    public static final boolean DEFAULT_EXTRAS = false;
    public static final boolean DEFAULT_AUTO_START = false;

    private static ModConfig SINGLE_INSTANCE = null;

    private final File configFile;

    private String username;
    private String oauthKey;
    private String prefix;
    private Color2 channelChatColor;
    private String timeFormatting;
    private boolean extras;
    private boolean autoStart;

    public ModConfig() {
        this.configFile = FabricLoader.getInstance().getConfigDir().resolve(MODID + ".json").toFile();
        this.username = DEFAULT_USERNAME;
        this.oauthKey = DEFAULT_OAUTH_KEY;
        this.prefix = DEFAULT_PREFIX;
        this.channelChatColor = DEFAULT_CHANNEL_CHAT_COLOR;
        this.timeFormatting = DEFAULT_TIME_FORMATTING;
        this.extras = DEFAULT_EXTRAS;
        this.autoStart = DEFAULT_AUTO_START;
    }

    public static ModConfig getConfig() {
        if (SINGLE_INSTANCE == null) {
            SINGLE_INSTANCE = new ModConfig();
        }

        return SINGLE_INSTANCE;
    }

    public void load() {
        try {
            String jsonStr = new String(Files.readAllBytes(this.configFile.toPath()));
            if (!jsonStr.equals("")) {
                JsonObject jsonObject = (JsonObject) JsonParser.parseString(jsonStr);
                this.username = jsonObject.has("username") ? jsonObject.getAsJsonPrimitive("username").getAsString() : DEFAULT_USERNAME;
                this.oauthKey = jsonObject.has("oauthKey") ? jsonObject.getAsJsonPrimitive("oauthKey").getAsString() : DEFAULT_OAUTH_KEY;
                this.prefix = jsonObject.has("prefix") ? jsonObject.getAsJsonPrimitive("prefix").getAsString() : DEFAULT_PREFIX;
                this.channelChatColor = jsonObject.has("channelChatColor") ? Color2.valueOf((jsonObject.getAsJsonPrimitive("channelChatColor").getAsString())) : DEFAULT_CHANNEL_CHAT_COLOR;
                this.timeFormatting = jsonObject.has("timeFormatting") ? jsonObject.getAsJsonPrimitive("timeFormatting").getAsString() : DEFAULT_TIME_FORMATTING;
                this.extras = jsonObject.has("extras") ? jsonObject.getAsJsonPrimitive("extras").getAsBoolean() : DEFAULT_EXTRAS;
                this.autoStart = jsonObject.has("autoStart") ? jsonObject.getAsJsonPrimitive("autoStart").getAsBoolean() : DEFAULT_AUTO_START;
            }
        } catch (IOException e) {
            // Do nothing, we have no file and thus we have to keep everything as default
        }
    }

    public void save() {
        JsonObject jsonObject = new JsonObject();
        jsonObject.addProperty("username", this.username);
        jsonObject.addProperty("oauthKey", this.oauthKey);
        jsonObject.addProperty("prefix", this.prefix);
        jsonObject.addProperty("channelChatColor", this.channelChatColor.getName());
        jsonObject.addProperty("timeFormatting", this.timeFormatting);
        jsonObject.addProperty("extras", this.extras);
        jsonObject.addProperty("autoStart", this.autoStart);

        try (PrintWriter out = new PrintWriter(configFile)) {
            out.println(jsonObject);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }


    public String getUsername() {
        return username;
    }
    public void setUsername(String username) {
        this.username = username;
    }

    public String getOauthKey() {
        return oauthKey;
    }
    public void setOauthKey(String oauthKey) {
        this.oauthKey = oauthKey;
    }

    public String getPrefix() {
        return prefix;
    }
    public void setPrefix(String prefix) {
        this.prefix = prefix;
    }

    public Color2 getChannelChatColor() {return channelChatColor;}
    public void setChannelChatColor(Color2 channelChatColor) {
        this.channelChatColor = channelChatColor;
    }

    public String getTimeFormatting() {
        return timeFormatting;
    }
    public void setTimeFormatting(String timeFormatting) {
        this.timeFormatting = timeFormatting;
    }
    public boolean getExtras() {
        return extras;
    }
    public void setExtras(boolean extras) {
        this.extras = extras;
    }
    public boolean getAutoStart() {return autoStart;}
    public void setAutoStart(boolean autoStart) {this.autoStart = autoStart;}
}

File: src/main/kotlin/com/theendercore/hydra/config/ModConfig.kt
package com.theendercore.hydra.config

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.theendercore.hydra.HydraMod.Companion.MODID
import net.fabricmc.loader.api.FabricLoader
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.PrintWriter
import java.nio.file.Files

class ModConfig {
    private val configFile: File = FabricLoader.getInstance().configDir.resolve("$MODID.json").toFile()
    var username: String
    var oauthKey: String
    var prefix: String
    var channelChatColor: Color?
    var timeFormatting: String
    var extras: Boolean
    var autoStart: Boolean
    var enableCache: Boolean

    init {
        username = DEFAULT_USERNAME
        oauthKey = DEFAULT_OAUTH_KEY
        prefix = DEFAULT_PREFIX
        channelChatColor = DEFAULT_CHANNEL_CHAT_COLOR
        timeFormatting = DEFAULT_TIME_FORMATTING
        extras = DEFAULT_EXTRAS
        autoStart = DEFAULT_AUTO_START
        enableCache = DEFAULT_ENABLE_CACHE
    }

    fun load() {
        try {
            val jsonStr = String(Files.readAllBytes(configFile.toPath()))
            if (jsonStr != "") {
                val jsonObject = JsonParser.parseString(jsonStr) as JsonObject
                username =
                    if (jsonObject.has("username")) jsonObject.getAsJsonPrimitive("username").asString else DEFAULT_USERNAME
                oauthKey =
                    if (jsonObject.has("oauthKey")) jsonObject.getAsJsonPrimitive("oauthKey").asString else DEFAULT_OAUTH_KEY
                prefix =
                    if (jsonObject.has("prefix")) jsonObject.getAsJsonPrimitive("prefix").asString else DEFAULT_PREFIX
                channelChatColor = if (jsonObject.has("channelChatColor")) Color.valueOf(
                    jsonObject.getAsJsonPrimitive("channelChatColor").asString
                ) else DEFAULT_CHANNEL_CHAT_COLOR
                timeFormatting =
                    if (jsonObject.has("timeFormatting")) jsonObject.getAsJsonPrimitive("timeFormatting").asString else DEFAULT_TIME_FORMATTING
                extras =
                    if (jsonObject.has("extras")) jsonObject.getAsJsonPrimitive("extras").asBoolean else DEFAULT_EXTRAS
                autoStart =
                    if (jsonObject.has("autoStart")) jsonObject.getAsJsonPrimitive("autoStart").asBoolean else DEFAULT_AUTO_START
                autoStart =
                    if (jsonObject.has("enableCache")) jsonObject.getAsJsonPrimitive("enableCache").asBoolean else DEFAULT_ENABLE_CACHE
            }
        } catch (e: IOException) {
            // Do nothing, we have no file,
        }
    }

    fun save() {
        val jsonObject = JsonObject()
        jsonObject.addProperty("username", username)
        jsonObject.addProperty("oauthKey", oauthKey)
        jsonObject.addProperty("prefix", prefix)
        jsonObject.addProperty("channelChatColor", channelChatColor?.getLabel())
        jsonObject.addProperty("timeFormatting", timeFormatting)
        jsonObject.addProperty("extras", extras)
        jsonObject.addProperty("autoStart", autoStart)
        jsonObject.addProperty("enableCache", enableCache)
        try {
            PrintWriter(configFile).use { out -> out.println(jsonObject) }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }
    }

    companion object {
        const val DEFAULT_USERNAME = ""
        const val DEFAULT_OAUTH_KEY = ""
        const val DEFAULT_PREFIX = "!"
        val DEFAULT_CHANNEL_CHAT_COLOR = Color.BLUE
        const val DEFAULT_TIME_FORMATTING = "HH:mm"
        const val DEFAULT_EXTRAS = false
        const val DEFAULT_AUTO_START = false
        const val DEFAULT_ENABLE_CACHE = false

        private var SINGLE_INSTANCE: ModConfig? = null
        val config: ModConfig?
            get() {
                if (SINGLE_INSTANCE == null) {
                    SINGLE_INSTANCE = ModConfig()
                }
                return SINGLE_INSTANCE
            }
    }
}